### PR TITLE
ACR auth simplification

### LINKED ISSUE #33820: [ACR] Optimize auth flow
ACR uses oauth2 and gets refresh token and access token.
A refresh token is obtained with AAD token and cached in memory. It's the same for the repository.

We get access token on each request using refresh token. The access token is scoped to specific actions and repository. It's valid (currently) for about an hour.

It's safe and doable to cache access token (either per action + client instance or bluntly once) and only refresh it on 401.

this is not blocking GA.

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes #33820

Reuses more logic from Bearer policy in core and does most of #33175 (except AcceessTokenCacheImpl which will come in #30581)

### CODE REVIEW

File: sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java [hunk 1]
```diff
@@ -170,7 +171,8 @@ private RequestBody toOkHttpRequestBody(BinaryData bodyContent, HttpHeaders head
 
         if (content instanceof ByteArrayContent
             || content instanceof StringContent
-            || content instanceof SerializableContent) {
+            || content instanceof SerializableContent
+            || content instanceof ByteBufferContent) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java:175]: Good catch, long term I think we need to add some more metadata to `BinaryDataContent` where we don't need to do all this type checking as it breaks when new types are added.

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java [hunk 2]
```diff
@@ -155,29 +152,35 @@ public static HttpPipeline buildHttpPipeline(
         HttpLoggingPolicy loggingPolicy = new HttpLoggingPolicy(logOptions);
         ArrayList<HttpPipelinePolicy> credentialPolicies = clone(policies);
         HttpPolicyProviders.addAfterRetryPolicies(credentialPolicies);
+        credentialPolicies.addAll(perRetryPolicies);
```
Comment (alzimmermsft) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java:155]: Could we make separate methods for building the credential pipeline and the request pipeline. Having both together could lead to an easy mistake that breaks stuff.

File: sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java (hunk 1 above)
Comment (lmolkova) [reply sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java:175]: reading my thoughts - https://github.com/Azure/azure-sdk-for-java/issues/34501

**Review by alzimmermsft — APPROVED** [review]

**Pull request merged**